src/waltz/xdp/fd_xdp1.c

```diff
@@ -32,56 +33,183 @@ struct __attribute__((aligned(8))) bpf_link_create {
                     ushort const * ports,
-                    ulong          ports_cnt ) {
+                    ulong          ports_cnt,
+                    int            allowed_gre ) {
+  
+  #define LBL_PASS      1   // Pass the packet to the kernel
+  #define LBL_REDIRECT  2   // Redirect the packet to firedancer software
+
+  #define LBL_GRE_CHECK     3   // Ethernet-->IPV4-->GRE-->Inner IPV4-->UDP
+  #define LBL_UDP_CHECK     4   // Ethernet-->IPV4-->UDP
+
   if( FD_UNLIKELY( ports_cnt>16UL ) ) {
     FD_LOG_ERR(( "Too many XDP UDP ports (%lu)", ports_cnt ));
   }
 
   ulong * code = code_buf;
-  *(code++) = FD_EBPF( ldxw, r2, r1, 0         );  // r2 = xdp_md->data
-  *(code++) = FD_EBPF( ldxw, r3, r1, 4         );  // r3 = xdp_md->data_end
-  *(code++) = FD_EBPF( mov64_reg, r4, r2       );
-  *(code++) = FD_EBPF( add64_imm, r4, 42       );
-  *(code++) = FD_EBPF( jgt_reg, r4, r3, +1     );  // if r2+42 > r3 goto lbl_pass
-  *(code++) = FD_EBPF( ldxh, r5, r2, 12        );
-  *(code++) = FD_EBPF( jne_imm, r5, 0x0008, +1 );  // if eth_hdr->net_type != IP4 goto lbl_pass
+  *(code++) = FD_EBPF( ldxw, r2, r1, 0                          );  // r2 = xdp_md->data
+  *(code++) = FD_EBPF( ldxw, r3, r1, 4                          );  // r3 = xdp_md->data_end
+
+  *(code++) = FD_EBPF( mov64_reg, r5, r2                        );
+  *(code++) = FD_EBPF( add64_imm, r5, 34                        );  // Bound check accessing the eth_hdr (14 bytes) and the ip4_hdr (20 bytes)
+  *(code++) = FD_EBPF( jgt_reg, r5, r3, LBL_PASS                );  // if r2+34 > r3 goto LBL_PASS
+
+  *(code++) = FD_EBPF( ldxh, r5, r2, 12                         );
+  *(code++) = FD_EBPF( jne_imm, r5, 0x0008, LBL_PASS            );  // if eth_hdr->net_type != IP4 goto LBL_PASS
+
+  /* Advance r2 to the start of first ip4_hdr */
+  *(code++) = FD_EBPF( add64_imm, r2, 14                        );  
+ 
+  /* Calculate the start of next hdr and store in r4 */
+  *(code++) = FD_EBPF( ldxb, r4, r2, 0                          );  // r4 = ip4_hdr->verihl
+  *(code++) = FD_EBPF( and64_imm, r4, 0x0f                      );  // r4 = ip4_hdr->ihl (lsb of ip4_hrd->verihl)
+  *(code++) = FD_EBPF( lsh64_imm, r4, 2                         );  // r4 = ip4_hdr->ihl*4 (length of ipv4 header)
+  *(code++) = FD_EBPF( add64_reg, r4, r2                        );  // r4 = &ip4_hdr + length of ip4_hdr = start of next hdr
+
+  /* Check if the next hdr is udp or gre */
+  *(code++) = FD_EBPF( ldxb, r5, r2, 9                          );  // r5 = ip4_hdr->protocol
+
+  if( allowed_gre ) {
+    *(code++) = FD_EBPF( jeq_imm, r5, 47, LBL_GRE_CHECK );  // if ip4_hdr->protocol == GRE goto gre_check
+  }
+
+  *(code++) = FD_EBPF( jeq_imm, r5, 17, LBL_UDP_CHECK               );  // if ip4_hdr->protocol == UDP goto udp_check
+  *(code++) = FD_EBPF( mov64_imm, r5, 0                         );  // r5 = 0
+  *(code++) = FD_EBPF( jeq_imm, r5, 0, LBL_PASS                 );  // goto LBL_PASS
```

Comment:
This is a bit odd ... This should be an unconditional jump (ja instruction). Not sure if our assembly generator knows that instruction though